src/workspace/mod.rs

```diff
@@ -209,17 +213,74 @@ impl Workspace {
         }
     }
 
+    fn supports_soft_kill(&self, winid: u32) -> bool {
+        let atom_id = get_internal_atom(&self.connection, Atom::WmProtocols.as_ref());
+
+        self.connection.flush().unwrap();
+        let atom_reply = self
+            .connection
+            .get_property(false, winid, atom_id, AtomEnum::ANY, 0, 1024)
+            .unwrap()
+            .reply();
+
+        if let Ok(atom_reply) = atom_reply {
+            self.connection.flush().unwrap();
+
+            let prop_type = match atom_reply.type_ {
+                0 => return false, // Null response
+                atomid => atom_name(&self.connection, atomid),
+            };
+
+            if prop_type == "ATOM" {
+                let atoms = atom_reply
+                    .value32()
+                    .unwrap()
+                    .map(|a| atom_name(&self.connection, a))
+                    .collect::<Vec<String>>();
+
+                return atoms.iter().any(|p| p == "WM_DELETE_WINDOW");
+            }
+        }
+        false
+    }
+
     pub fn kill_window(&mut self, winid: &u32) {
-        //TODO implement soft kill via client message over x
-        //(Tell window to close itself)
-        //https://github.com/DHBW-FN/OxideWM/issues/46
-        self.remove_window(winid);
-        self.connection
-            .kill_client(*winid)
-            .expect("Could not kill client");
+        if self.supports_soft_kill(*winid) {
+            self.execute_softkill(*winid);
+        } else {
+            self.remove_window(winid);
+            self.connection
+                .kill_client(*winid)
+                .expect("Could not kill client");
```

Comment:
Usually I'd first try to kill the client and than remove the window if something were to happen but I guess since we panic if that were to happen... ?:)

Reply:
I agree to this. I guess it is unlikely to happen but just in case this seems to be the better way :)

Reply:
i dont think it makes a difference, in a case where this fails something went really wrong somehow..
but i will change the order.